File: p003_medium_longest_substring_without_repeating_characters.java
import java.util.HashMap;
import java.util.Map;

// class Solution
// using Array will be even faster (2x)
class longestSubstringWithoutRepeatingCharacters {
    public int lengthOfLongestSubstring(String s) {
        int maxResult = 0;
        int currResult = 0;
        int start = 0;
        Map<Character, Integer> map = new HashMap<>();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (!map.containsKey(c) || map.get(c) < start) {
                currResult++;
                maxResult = Math.max(currResult, maxResult);
            } else {
                start = map.get(c) + 1;
                currResult = i - start + 1;
            }
            map.put(c, i);
        }
        return maxResult;
    }
}

File: p003_medium_longest_substring_without_repeating_characters.py
# str.index (sub, start=None, end=None)
# ValueError: substring not found or lowest index
#

# str.find (sub, start=None, end=None) -1 or lowest index
# for highest index: rindex rfind

class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        # if every time not found: O(n^2)
        if s == "":
            return 0
        curr_result = max_result = 1
        start = 0
        end = 1
        for i, v in enumerate(s):
            if i == 0:
                continue
            h_index = s.rfind(v, start, end)
            if h_index == -1:
                # -1: add this one
                curr_result += 1
                max_result = max(curr_result, max_result)
            else:
                start = h_index + 1
                curr_result = end - h_index
            end += 1
        return max_result
